src/mainframe/database.py

```diff
@@ -21,3 +27,80 @@ def get_db() -> Generator[Session, None, None]:
         yield session
     finally:
         session.close()
+
+
+class StorageProtocol(Protocol):
+    def lookup_packages(
+        self, name: Optional[str] = None, version: Optional[str] = None, since: Optional[dt.datetime] = None
+    ) -> Sequence[Scan]:
+        """
+        Lookup information on scanned packages based on name, version, or time
+        scanned. If multiple packages are returned, they are ordered with the most
+        recently queued package first.
+
+        Args:
+            since: A int representing a Unix timestamp representing when to begin the search from.
+            name: The name of the package.
+            version: The version of the package.
+            session: DB session.
+
+        Exceptions:
+            ValueError: Invalid parameter combination was passed. See below.
+
+        Returns:
+            Sequence of `Scan`s, representing the results of the query
+
+        Only certain combinations of parameters are allowed. A query is valid if any of the following combinations are used:
+            - `name` and `version`: Return the package with name `name` and version `version`, if it exists.
+            - `name` and `since`: Find all packages with name `name` since `since`.
+            - `since`: Find all packages since `since`.
+            - `name`: Find all packages with name `name`.
+        All other combinations are disallowed.
+
+        In more formal terms, a query is valid
+            iff `((name and not since) or (not version and since))`
+        where a given variable name means that query parameter was passed. Equivalently, a request is invalid
+            iff `(not (name or since) or (version and since))`
+        """
+        ...
+
+    def mark_reported(self, *, scan: Scan, subject: str) -> None:
+        """Mark the given `Scan` record as reported by `subject`."""
+        ...
+
+
+class DatabaseStorage(StorageProtocol):
+    def __init__(self, sessionmaker: orm.sessionmaker[Session]):
+        self.sessionmaker = sessionmaker
```

Comment:
Each instance should have one session that commits changes when the `DatabaseStorage` gets destroyed. This is the 'Unit of Work" pattern.

Reply:
I originally thought of using one `DatabaseStorage` instance across the whole program as a singleton, though we can also create and destroy them for each endpoint like you're suggesting. What are the advantages to this method, rather than having each individual method of this class manage it's own session and unit of work?